torch/_higher_order_ops/flex_attention.py

```diff
@@ -69,6 +69,74 @@ def _permute_strides(out: torch.Tensor, query_strides: Tuple[int, ...]) -> torch
     return new_out
 
 
+@torch.library.custom_op("mylib::zeros_and_scatter", mutates_args=())
+def zeros_and_scatter(
+    shape: List[int],
+    indices: List[Tensor],
+    vals: Tensor,
+) -> Tensor:
+    grad = torch.zeros(shape, device=vals.device, dtype=vals.dtype)
+
+    for i, idx in enumerate(indices):
+        assert isinstance(idx, torch.Tensor)  # Appease the mypy overlords
+        indices[i] = idx.expand(vals.shape)
+
+    return torch.ops.aten.index_put(grad, indices, vals, accumulate=True)
+
+
+@zeros_and_scatter.register_fake
+def _(
+    shape: List[int],
+    indices: List[Tensor],
+    vals: Tensor,
+) -> Tensor:
+    return torch.empty(shape, device=vals.device)
+
+
+@zeros_and_scatter.register_vmap
+def _(info, in_dims, shape, indices, val):
+    if len(indices) > 1:
+        for i, idx in enumerate(indices):
+            if torch._C._functorch.is_batchedtensor(idx):
+                indices[i] = idx.unsqueeze(-1)
```

Comment:
the batching rule shouldn't need to contain APIs like `torch._C._functorch.is_batchedtensor` -- it should be written to be BatchedTensor agnostic. But if you're on a timeline feel free to figure this out later